docs/proposals/node-local-services.md

```diff
@@ -0,0 +1,269 @@
+interfaces on the node. Should there be concerns about the cost of looking up
+all the interfaces on a regular basis, a simple one-entry cache (an `<address,
+interface>` pair, with the full search as a fallback) might obviate a lot of the
+extra overhead.
+
+#### iptables kernel mode
+
+In the existing implementation, `kube-proxy` keeps track of all N endpoints for
+a service. On each node, it creates `2*N` iptables rules to pick a random
+endpoint when opening a new connection: one rule for the probabilistic selection
+and another one for the actual `DNAT` packet rewriting. For node-local services,
+it will require only one rule or even zero, if it keeps track of the actual pod
+liveness. The end result is less overhead over a conventional service, before
+even taking into account network latency.
+
+For use cases like `kube2iam`, which requires intercepting traffic meant for the
+special `169.254.169.254` metadata IP address, perhaps `ClusterIP` should be
+allowed to live outside of the service IP range. For security reasons, this
+behaviour would require explicit activation from the administrator (an API
+server or proxy flag).
+
+#### ipvs kernel mode
+
+Issue [#17470](https://github.com/kubernetes/kubernetes/issues/17470) calls for
+a newer ipvs mode. This would work similarly to iptables mode.
+
+#### proxy userland mode
+
+In this mode, iptables rules currently redirect the VIP to a local port owned by
+the proxy, which then forwards traffic back and forth between the client and the
+server pod. For node-local services, there is no need to make traffic go through
+the proxy, since there are no endpoint candidates to choose from. Thus, the code
+will be very similar to, if not shared with, the iptables case.
+
+### n-to-1 forwarding
+
+For scenarios like `loasd`, it would be useful to let multiple services point to
+the same daemon. Setting multiple `DNAT` rules translating from different VIPs
+to the same local port is trivial, but in most cases the daemon will want to
+find out which VIP a connection was meant for. Possible options:
+
+1. Have DaemonSets open N different ports, with node-local services pointing to
+   a specific Daemonset and port
+1. Rely on in-band, protocol-specific information, such as the `Host:` header
+   for HTTP traffic (which might make use of HTTPS more complicated)
+1. Forward traffic to addresses in the `127.0.0.0/8` network, while at the same
+   time annotating node-local services with their individual loopback address.
+   This also requires:
+   - keeping track of the _VIP_->_loopback_ translations in the controller
+     manager
+   - having all the node-local daemons talk to the API server to watch services
+   - enabling the routing of external traffic destined for the loopback address
+   range, by running `sysctl -w net.ipv4.conf.eth0.route_localnet=1` on each
+   node
+
+   and happens to be crazy (if it can be made to work reliably at all).
+
+## Implementation plan
```

Comment:
If we would only support DNS based lookups, couldnÄt this be solved on the node-level if the kubelet or some other component was a DNS resolver?
The node local DNS resolver would resolve a fqdn into a local address whenever the backing service is considered to be node-local.
Otherwise it would just dispatch the request to kube-dns.